semanticdb3/semanticdb3.md

```diff
@@ -0,0 +1,491 @@
+
+TODO: Once we upgrade semanticdb-scalac to support SemanticDB v3,
+demonstrate a hello world example. Show the original program, a prettyprinted
+version of its SemanticDB data, very briefly explain the sections.
+
+TODO: Once we ship metac and metap that support SemanticDB v3, explain that
+the SemanticDB data model is materialized as Protocol Buffers [\[13\]][13], JSON [\[14\]][14]
+and SQL [\[15\]][15]. Demonstrate how we can use metac to generate `.semanticdb` files
+from the hello world example above, show the binary payload, decode the files with metap.
+
+TODO: Explain how these protobuf files can be used in tools like scalafix, metadoc and metals.
+Mention that beyond protobuf files, there are many ways to use SemanticDB as an
+interchange format, including passing objects around in memory, communication via JSON, etc etc.
+
+## Model
+
+In this section, we describe the SemanticDB data model by going through the
+individual sections of the associated protobuf schema. However, in addition to protobuf,
+we also support other kinds of schemas, including JSON and SQL.
+See [Schemas](#schemas) for more information.
+
+### TextDocument
+
+```protobuf
+message TextDocuments {
+  repeated TextDocument textDocuments = 1;
+}
+
+message TextDocument {
+  string model = 255;
+  string uri = 1;
+  string text = 2;
+  string language = 3;
+  repeated Identifier identifiers = 4;
+  repeated SymbolInformation symbols = 5;
+  repeated Diagnostic diagnostics = 6;
+  repeated Synthetic synthetics = 7;
+}
+```
+
+`TextDocument` is the central data structure in the SemanticDB model.
+It provides semantic information about a snippet of code written in a programming
+language `language`.
+
+SemanticDB payloads must include the descriptor of the data model that was used
+to produce them in the `model` field. The following descriptors are supported:
+
+<table>
+  <tr>
+    <td><b>Descriptor</b></td>
+    <td><b>Explanation</b></td>
+    <td><b>Data model<b></td>
+  </tr>
+  <tr>
+    <td>&lt;missing&gt;</td>
+    <td>Legacy SemanticDB payloads</td>
+    <td><a href="https://github.com/scalameta/scalameta/blob/v2.1.5/langmeta/langmeta/shared/src/main/protobuf/semanticdb.proto">semanticdb.proto</a></td>
+  </tr>
+  <tr>
+    <td><code>semanticdb3</code></td>
+    <td>Current version of the model</td>
+    <td>Described in this document</td>
+  </tr>
+</table>
+
+Snippets of code referenced in `TextDocument` can be defined
+in one of three ways: 1) via a URI [\[16\]][16] provided in `uri`,
+2) via a string provided in `text`, 3) via a combination of both.
+
+Semantic information is stored in so called sections - repeated fields within
+the message definition - as described below. These sections are optional, which
+means that documents providing only part of semantic information for the corresponding snippet
+(or no semantic information at all) are completely legal.
+
+### Range
+
+```protobuf
+message Range {
+  int32 start_line = 1;
```

Comment:
I still have doubts about this change, I believe start/end offsets are a superior format and I have concerns about performance implications. For tooling you will always need to have the buffer available anyways, in which case converting between the two formats is easy.

Reply:
Fair enough, Index-While-Building seems to use line/column as well https://github.com/apple/swift-clang/blob/stable/test/Index/Store/Inputs/json.c.json

> Also, why does one always need to have the buffer available anyways?

Because you need the ability to account for minor changes in the file contents. Blindly returning reference positions may result in incorrect rename refactorings or bogus highlighted code regions. If you have something like sha256 it's better, but then a single character insert makes the entire snapshot unusable.

Reply:
the readability of the raw data is improved with line/column, but readability will go down the drain anyways if we use symbol lookup tables to compact the data.

Reply:
Metals btw already has support to recover from slightly stale semanticdb documents using token edit-distance, and that requires the full text contents. In this space, I think we are actually ahead of alternative technologies and the lesson learned is that you need the text contents anyways so you might as well use start/end.

Reply:
Regardless, I'm fine with using line/character for the simplicity of manually constructing and validating raw data. I would personally prefer `column` over `character` since I think it's confusing to use `character` like is done in LSP. In my editor and in clang they use `col` 

![screen shot 2018-01-19 at 19 13 53](https://user-images.githubusercontent.com/1408093/35165018-f99acafe-fd4c-11e7-80f9-5832eb26e285.png)

Reply:
Speaking of always having the document text, what about SQL? When you're working with the SQL index, you don't have the luxury of positions on the fly. I suspect that many large-scale code analysis/rewrite jobs will be similarly restrictive.

I think a combination of a fingerprint and optional client smartness (like the one that is implemented in Metals) is the way to go, but let's defer this discussion to a follow-up update to the spec. There are multiple questions in play here - do we fingerprint just the contents of the source file, or its dependencies as well? do we allow fingerprinting to be buildtool-specific? - and they all seem orthogonal to the goal of this pullrequest.

Reply:
Speaking of the choice between offset and line/column, let's put a hold on this until we have the benchmark data. Then we will be able to talk about readability, too.

Reply:
I have noticed the usability of raw offsets isn’t great and will admit to pasting hunks of code in an online character counter to quickly figure out what offset I wanted 😊 
On the other hand, offsets are great for dbs that support interval types, especially for fast overlaps lookups.  We could use a datatype that allows encoding two ints in one value, say a decimal type.

Reply:
Here's benchmark results: https://github.com/xeno-by/scalameta/commit/be54bca11542580f76b670870318ac238f997663. Looks like the effect of refactoring positions is within the fluctuation window.